File: 2025/GoalTracker/backend/src/main/java/com/goal/service/GoalService.java
package com.goal.service;

import com.goal.model.Goal;
import com.goal.repository.GoalRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class GoalService {

    @Autowired
    private GoalRepository goalRepository;

    public List<Goal> getAllGoals() {
        return goalRepository.findAll();
    }

    public Goal getGoalById(Long id) {
        Optional<Goal> goal = goalRepository.findById(id);
        return goal.orElse(null);
    }

    public List<Goal> searchGoals(String query) {
        return goalRepository.searchByTitleOrDescription(query);
    }

    public List<Goal> getGoalsByCategory(String category) {
        return goalRepository.findByCategory(category);
    }

    public List<Goal> getGoalsByStatus(Goal.Status status) {
        return goalRepository.findByStatus(status);
    }

    public List<Goal> getGoalsByPriority(Goal.Priority priority) {
        return goalRepository.findByPriority(priority);
    }

    public Goal createGoal(Goal goal) {
        return goalRepository.save(goal);
    }

    public Goal updateGoal(Long id, Goal goalDetails) {
        Optional<Goal> optionalGoal = goalRepository.findById(id);
        if (optionalGoal.isPresent()) {
            Goal goal = optionalGoal.get();
            goal.setTitle(goalDetails.getTitle());
            goal.setDescription(goalDetails.getDescription());
            goal.setCategory(goalDetails.getCategory());
            goal.setPriority(goalDetails.getPriority());
            goal.setStatus(goalDetails.getStatus());
            goal.setProgress(goalDetails.getProgress());
            goal.setTargetDate(goalDetails.getTargetDate());
            return goalRepository.save(goal);
        }
        return null;
    }

    public Goal updateProgress(Long id, Integer progress) {
        Optional<Goal> optionalGoal = goalRepository.findById(id);
        if (optionalGoal.isPresent()) {
            Goal goal = optionalGoal.get();
            goal.setProgress(progress);
            // Auto-update status based on progress
            if (progress >= 100 && goal.getStatus() != Goal.Status.COMPLETED) {
                goal.setStatus(Goal.Status.COMPLETED);
            } else if (progress > 0 && progress < 100 && goal.getStatus() == Goal.Status.ACTIVE) {
                goal.setStatus(Goal.Status.IN_PROGRESS);
            }
            return goalRepository.save(goal);
        }
        return null;
    }

    public Goal completeGoal(Long id) {
        Optional<Goal> optionalGoal = goalRepository.findById(id);
        if (optionalGoal.isPresent()) {
            Goal goal = optionalGoal.get();
            goal.setProgress(100);
            goal.setStatus(Goal.Status.COMPLETED);
            return goalRepository.save(goal);
        }
        return null;
    }

    public boolean deleteGoal(Long id) {
        if (goalRepository.existsById(id)) {
            goalRepository.deleteById(id);
            return true;
        }
        return false;
    }
}

File: 2025/GoalTracker/frontend/src/app/services/goal.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Goal } from '../models/goal';

@Injectable({
  providedIn: 'root'
})
export class GoalService {
  private apiUrl = 'http://localhost:8093/api/goals';

  constructor(private http: HttpClient) {}

  getAllGoals(): Observable<Goal[]> {
    return this.http.get<Goal[]>(this.apiUrl);
  }

  getGoalById(id: number): Observable<Goal> {
    return this.http.get<Goal>(`${this.apiUrl}/${id}`);
  }

  searchGoals(query: string): Observable<Goal[]> {
    const params = new HttpParams().set('query', query);
    return this.http.get<Goal[]>(`${this.apiUrl}/search`, { params });
  }

  getGoalsByStatus(status: string): Observable<Goal[]> {
    return this.http.get<Goal[]>(`${this.apiUrl}/status/${status}`);
  }

  getGoalsByCategory(category: string): Observable<Goal[]> {
    return this.http.get<Goal[]>(`${this.apiUrl}/category/${category}`);
  }

  getGoalsByPriority(priority: string): Observable<Goal[]> {
    return this.http.get<Goal[]>(`${this.apiUrl}/priority/${priority}`);
  }

  createGoal(goal: Goal): Observable<Goal> {
    return this.http.post<Goal>(this.apiUrl, goal);
  }

  updateGoal(id: number, goal: Goal): Observable<Goal> {
    return this.http.put<Goal>(`${this.apiUrl}/${id}`, goal);
  }

  updateProgress(id: number, progress: number): Observable<Goal> {
    return this.http.put<Goal>(`${this.apiUrl}/${id}/progress`, progress);
  }

  completeGoal(id: number): Observable<Goal> {
    return this.http.put<Goal>(`${this.apiUrl}/${id}/complete`, {});
  }

  deleteGoal(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}


